File: app/src/main/java/com/example/messengerandroid/service/CharacterRepository.java
package com.example.messengerandroid.service;

import androidx.annotation.NonNull;

import com.example.messengerandroid.model.Character;

import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class CharacterRepository {

    private final ApiService apiService;

    public CharacterRepository() {
        apiService = ApiClient.getClient().create(ApiService.class);
    }

    public void getCharacters(int page, int pageSize, CharacterCallback callback) {
        apiService.getCharacters(page, pageSize).enqueue(new Callback<List<Character>>() {
            @Override
            public void onResponse(@NonNull Call<List<Character>> call,
                                   @NonNull Response<List<Character>> response) {
                if (response.isSuccessful() && response.body() != null) {
                    callback.onSuccess(response.body());
                } else {
                    callback.onError(new Exception("Failed to fetch characters"));
                }
            }

            @Override
            public void onFailure(Call<List<Character>> call, Throwable t) {
                callback.onError(t);
            }
        });
    }

    public interface CharacterCallback {
        void onSuccess(List<Character> characters);
        void onError(Throwable t);
    }
}

File: app/src/main/java/com/example/messengerandroid/service/CharacterRepository.kt
package com.example.messengerandroid.service

import com.example.messengerandroid.model.Character;

class CharacterRepository {

    private val apiService: ApiService = ApiClient.getClient().create(ApiService::class.java)

    fun getCharacters(page: Int, pageSize: Int, callback: CharacterCallback) {
        apiService.getCharacters(page, pageSize).enqueue(object : retrofit2.Callback<List<Character>> {
            override fun onResponse(
                call: retrofit2.Call<List<Character>>,
                response: retrofit2.Response<List<Character>>
            ) {
                if (response.isSuccessful && response.body() != null) {
                    callback.onSuccess(response.body()!!)
                } else {
                    callback.onError(Exception("Failed to fetch characters"))
                }
            }

            override fun onFailure(call: retrofit2.Call<List<Character>>, t: Throwable) {
                callback.onError(t)
            }
        })
    }

    interface CharacterCallback {
        fun onSuccess(characters: List<Character>)
        fun onError(t: Throwable)
    }
}
